File: sample/state_space/discrete_state_space_1.py
"""
File: discrete_state_space_1.py

This script demonstrates the simulation and C++ code generation of a discrete-time state-space model using Python.
It defines a simple state-space system, simulates its response over a number of time steps, and visualizes the results.
Additionally, it generates C++ header files for the defined state-space model for deployment purposes.
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

sys.path.append(str(Path(__file__).resolve().parents[1]))
sys.path.append(str(Path(__file__).resolve().parents[2]))

import numpy as np
import control
import matplotlib.pyplot as plt

from python_control.state_space_deploy import StateSpaceDeploy
from sample.simulation_manager.visualize.simulation_plotter_dash import SimulationPlotterDash

# define state-space model
A = np.array([[0.7, 0.2],
              [-0.3, 0.8]])
B = np.array([[0.1],
              [0.2]])
C = np.array([[2, 0]])
D = np.array([[0]])

dt = 0.01
sys = control.ss(A, B, C, D, dt)

# You can create cpp header which can easily define state space as C++ code
deployed_file_names = StateSpaceDeploy.generate_state_space_cpp_code(
    sys, number_of_delay=0)
print(deployed_file_names)

# initialize state and input
x = np.array([[0],
              [0]])
u = 1  # input
n_steps = 50  # number of steps

plotter = SimulationPlotterDash()

# simulation
for _ in range(n_steps):
    y = C @ x + D * u
    x = A @ x + B * u

    plotter.append(x)
    plotter.append(y)

plotter.assign("x", column=0, row=0, position=(0, 0))
plotter.assign("x", column=1, row=0, position=(0, 0))
plotter.assign("y", column=0, row=0, position=(1, 0))

plotter.plot("Discrete-Time State-Space Response")

File: sample/state_space/discrete_state_space_1.cpp
/**
 * @file discrete_state_space_1.cpp
 * @brief Example simulation of a discrete-time state-space system in C++.
 *
 * This file demonstrates the creation and simulation of a discrete-time
 * state-space model. The state-space system is defined by matrices A, B, C, and
 * D, and is simulated over 50 time steps with a constant input. The state and
 * output values are printed at each simulation step.
 */
#include <iostream>

#include "python_control.hpp"
#include "python_numpy.hpp"

using namespace PythonNumpy;
using namespace PythonControl;

int main(void) {
  /* Define State Space */
  auto A = make_DenseMatrix<2, 2>(0.7, 0.2, -0.3, 0.8);
  auto B = make_DenseMatrix<2, 1>(0.1, 0.2);
  auto C = make_DenseMatrix<1, 2>(2.0, 0.0);
  auto D = make_DenseMatrix<1, 1>(0.0);
  double dt = 0.01;

  auto sys = make_DiscreteStateSpace(A, B, C, D, dt);

  /* State Space Simulation */
  for (std::size_t sim_step = 0; sim_step < 50; ++sim_step) {
    auto u = make_StateSpaceInput<1>(1.0);

    sys.update(u);

    std::cout << "X_0: " << sys.get_X()(0, 0) << ", ";
    std::cout << "X_1: " << sys.get_X()(1, 0) << ", ";
    std::cout << "Y: " << sys.get_Y()(0, 0) << ", ";
    std::cout << std::endl;
  }

  return 0;
}
